File: neo/backend/src/app/api/v1/agent_prototype.py
"""Agent Prototype API routes."""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.models.user import User
from app.schemas.agent_prototype import (
    AgentPrototypeCreate,
    AgentPrototypeListResponse,
    AgentPrototypePublish,
    AgentPrototypeResponse,
    AgentPrototypeRollback,
    AgentPrototypeStatusUpdate,
    AgentPrototypeUpdate,
    AgentPrototypeVersionListResponse,
    NextVersionResponse,
)
from app.schemas.response import ApiResponse
from app.services.agent_prototype_service import AgentPrototypeService

router = APIRouter(prefix="/agent_prototype", tags=["Agent Prototype"])


def get_service(db: Session = Depends(get_db)) -> AgentPrototypeService:
    """Get Agent Prototype service."""
    return AgentPrototypeService(db)


@router.get("", response_model=ApiResponse[AgentPrototypeListResponse])
def list_prototypes(
    service: AgentPrototypeService = Depends(get_service),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    status: str | None = Query(None, description="Filter by status: draft/enabled/disabled"),
    type: str | None = Query(None, description="Filter by type: site_operation/expert_interview"),
    search: str | None = Query(None, max_length=100, description="Search by name"),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeListResponse]:
    """List all Agent Prototypes with pagination and filters."""
    # Convert status string to enum
    from app.schemas.agent_prototype import AgentPrototypeStatus

    status_enum = None
    if status:
        try:
            status_enum = AgentPrototypeStatus(status)
        except ValueError:
            pass

    prototypes, total = service.list_prototypes(
        status=status_enum,
        agent_type=type,  # type is the API param, maps to agent_type in repo
        search=search,
        page=page,
        page_size=page_size,
    )

    return ApiResponse.success(
        AgentPrototypeListResponse(
            items=[AgentPrototypeResponse.model_validate(p) for p in prototypes],
            total=total,
            page=page,
            page_size=page_size,
        ),
    )


@router.post("", response_model=ApiResponse[AgentPrototypeResponse])
def create_prototype(
    data: AgentPrototypeCreate,
    service: AgentPrototypeService = Depends(get_service),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Create a new Agent Prototype."""
    prototype = service.create_prototype(data, user_id=int(current_user.id))
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))


@router.get("/next-version/{prototype_id}", response_model=ApiResponse[NextVersionResponse])
def get_next_version(
    prototype_id: int,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[NextVersionResponse]:
    """Get the next version number for a prototype."""
    next_version = service.get_next_version(prototype_id)
    return ApiResponse.success(NextVersionResponse(next_version=next_version))


@router.get("/{prototype_id}", response_model=ApiResponse[AgentPrototypeResponse])
def get_prototype(
    prototype_id: int,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Get an Agent Prototype by ID."""
    prototype = service.get_prototype(prototype_id)
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))


@router.put("/{prototype_id}", response_model=ApiResponse[AgentPrototypeResponse])
def update_prototype(
    prototype_id: int,
    data: AgentPrototypeUpdate,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Update an Agent Prototype."""
    prototype = service.update_prototype(prototype_id, data)
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))


@router.delete("/{prototype_id}", response_model=ApiResponse[dict])
def delete_prototype(
    prototype_id: int,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[dict]:
    """Delete an Agent Prototype (only draft status)."""
    service.delete_prototype(prototype_id)
    return ApiResponse.success({"message": "Prototype deleted successfully"})


@router.put("/{prototype_id}/status", response_model=ApiResponse[AgentPrototypeResponse])
def update_status(
    prototype_id: int,
    data: AgentPrototypeStatusUpdate,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Update Agent Prototype status (enable/disable)."""
    prototype = service.update_status(prototype_id, data)
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))


@router.post("/{prototype_id}/publish", response_model=ApiResponse[AgentPrototypeResponse])
def publish_prototype(
    prototype_id: int,
    data: AgentPrototypePublish,
    service: AgentPrototypeService = Depends(get_service),
    current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Publish a new version of the Agent Prototype."""
    prototype = service.publish_prototype(prototype_id, data, user_id=current_user.id)
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))


@router.get("/{prototype_id}/versions", response_model=ApiResponse[AgentPrototypeVersionListResponse])
def get_version_history(
    prototype_id: int,
    service: AgentPrototypeService = Depends(get_service),
    _current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeVersionListResponse]:
    """Get version history for a prototype."""
    from app.schemas.agent_prototype import AgentPrototypeVersionResponse

    versions = service.get_version_history(prototype_id)
    return ApiResponse.success(
        AgentPrototypeVersionListResponse(
            items=[AgentPrototypeVersionResponse.model_validate(v) for v in versions],
            total=len(versions),
        ),
    )


@router.post("/{prototype_id}/rollback", response_model=ApiResponse[AgentPrototypeResponse])
def rollback_prototype(
    prototype_id: int,
    data: AgentPrototypeRollback,
    service: AgentPrototypeService = Depends(get_service),
    current_user: dict = Depends(get_current_user),
) -> ApiResponse[AgentPrototypeResponse]:
    """Rollback Agent Prototype to a specific version."""
    prototype = service.rollback_prototype(prototype_id, data, user_id=current_user.id)
    return ApiResponse.success(AgentPrototypeResponse.model_validate(prototype))

File: neo/frontend/lib/api/agent-prototype.ts
/**
 * Agent Prototype API client
 * Connects to backend API via frontend proxy: /api/v1/agent_prototype
 */

import type {
  AgentPrototypeStatus,
  PromptConfig,
} from "@/components/agent-prototype/agent-prototype-types";
import { getAuthHeaders } from "@/lib/utils/auth";

// API base URL - use frontend proxy in production
const API_BASE = "/api/v1";

// Types matching backend API
// Agent type enum
export type AgentPrototypeType = "site_operation" | "expert_interview";

export interface AgentPrototypeResponse {
  id: number;
  code: string;
  name: string;
  description: string | null;
  version: string | null;
  model: string; // Legacy field, kept for backward compatibility
  provider_id: number | null;
  model_id: string | null;
  llm_config: Record<string, unknown> | null;
  prompts: PromptConfig;
  config: Record<string, unknown>;
  status: AgentPrototypeStatus;
  type: AgentPrototypeType;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface AgentPrototypeListResponse {
  items: AgentPrototypeResponse[];
  total: number;
  page: number;
  page_size: number;
}

export interface AgentPrototypeVersionResponse {
  id: number;
  version: string;
  change_summary: string | null;
  is_rollback: boolean;
  created_by: number;
  created_at: string;
}

export interface AgentPrototypeVersionListResponse {
  items: AgentPrototypeVersionResponse[];
  total: number;
}

export interface CreateAgentPrototypeRequest {
  name: string;
  code?: string;
  description?: string;
  model?: string; // Legacy field
  provider_id?: number;
  model_id?: string;
  llm_config?: Record<string, unknown>;
  prompts?: PromptConfig;
  config?: Record<string, unknown>;
  type?: AgentPrototypeType;
}

export interface UpdateAgentPrototypeRequest {
  name?: string;
  description?: string;
  model?: string; // Legacy field
  provider_id?: number;
  model_id?: string;
  llm_config?: Record<string, unknown>;
  prompts?: PromptConfig;
  config?: Record<string, unknown>;
  type?: AgentPrototypeType;
}

export interface PublishAgentPrototypeRequest {
  change_summary: string;
}

export interface RollbackAgentPrototypeRequest {
  target_version_id: number;
}

export interface UpdateStatusRequest {
  status: AgentPrototypeStatus;
}

// Error handling
export class ApiError extends Error {
  constructor(
    message: string,
    public code: number,
    public statusCode: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

// Helper function for API requests
async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const url = `${API_BASE}${endpoint}`;
  const authHeaders = getAuthHeaders();

  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...authHeaders,
      ...options.headers,
    },
    credentials: "include",
  });

  const data = await response.json();

  if (!response.ok) {
    throw new ApiError(
      data.message || "Request failed",
      data.code || response.status,
      response.status,
    );
  }

  if (data.code !== 0) {
    throw new ApiError(data.message, data.code, 200);
  }

  return data.data as T;
}

// ==================== Agent Prototype API ====================

/**
 * List all agent prototypes with pagination and filters
 */
export async function listAgentPrototypes(params?: {
  page?: number;
  page_size?: number;
  status?: AgentPrototypeStatus | "all";
  search?: string;
}): Promise<AgentPrototypeListResponse> {
  const searchParams = new URLSearchParams();

  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.page_size)
    searchParams.set("page_size", params.page_size.toString());
  if (params?.status && params.status !== "all") {
    searchParams.set("status", params.status);
  }
  if (params?.search) searchParams.set("search", params.search);

  const queryString = searchParams.toString();
  const endpoint = `/agent_prototype${queryString ? `?${queryString}` : ""}`;

  return apiRequest<AgentPrototypeListResponse>(endpoint);
}

/**
 * Get a single agent prototype by ID
 */
export async function getAgentPrototype(
  prototypeId: number,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>(`/agent_prototype/${prototypeId}`);
}

/**
 * Create a new agent prototype
 */
export async function createAgentPrototype(
  data: CreateAgentPrototypeRequest,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>("/agent_prototype", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * Update an agent prototype
 */
export async function updateAgentPrototype(
  prototypeId: number,
  data: UpdateAgentPrototypeRequest,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>(`/agent_prototype/${prototypeId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * Delete an agent prototype (only draft status allowed)
 */
export async function deleteAgentPrototype(
  prototypeId: number,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(`/agent_prototype/${prototypeId}`, {
    method: "DELETE",
  });
}

/**
 * Publish an agent prototype with a new version
 */
export async function publishAgentPrototype(
  prototypeId: number,
  data: PublishAgentPrototypeRequest,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>(
    `/agent_prototype/${prototypeId}/publish`,
    {
      method: "POST",
      body: JSON.stringify(data),
    },
  );
}

/**
 * Get next version number for a prototype
 */
export async function getNextVersion(
  prototypeId: number,
): Promise<{ next_version: string }> {
  return apiRequest<{ next_version: string }>(
    `/agent_prototype/next-version/${prototypeId}`,
  );
}

/**
 * Get version history for a prototype
 */
export async function getVersionHistory(
  prototypeId: number,
): Promise<AgentPrototypeVersionListResponse> {
  return apiRequest<AgentPrototypeVersionListResponse>(
    `/agent_prototype/${prototypeId}/versions`,
  );
}

/**
 * Rollback to a specific version
 */
export async function rollbackAgentPrototype(
  prototypeId: number,
  data: RollbackAgentPrototypeRequest,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>(
    `/agent_prototype/${prototypeId}/rollback`,
    {
      method: "POST",
      body: JSON.stringify(data),
    },
  );
}

/**
 * Update agent prototype status (enable/disable)
 */
export async function updateAgentPrototypeStatus(
  prototypeId: number,
  data: UpdateStatusRequest,
): Promise<AgentPrototypeResponse> {
  return apiRequest<AgentPrototypeResponse>(
    `/agent_prototype/${prototypeId}/status`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    },
  );
}
